### PR TITLE
Bugfix/optimize likelihood

### LINKED ISSUE #46: Optimize likelihood_ratio
Optimize the likelihood_ratio calculation in the main loop, which currently takes up about 10-20% of the runtime (depending on the run). Verify that any optimizations made don't affect the output's correctness via unit and integration testing.

### PR DESCRIPTION
**Pull request by JHDeerin**
Improved `likelihood_ratios` calculation speed by ~600% (by eliminating a for loop + using a more optimized version of `nanmeans`). Since `likelihood_ratios` previously took up ~15% of the total runtime, this should improve overall algorithm performance by about 12.5%. Closes #46.

Adds 1 new Python dependency, a library called [bottleneck](https://bottleneck.readthedocs.io/en/latest/).

## Further improvements

There's still 1 for loop remaining that I'm not sure how to get rid of (the iteration over the loci). After that, examining the code for more complicated algorithmic optimizations and/or verifying there aren't existing numpy functions that can combine some of this could result in further speed improvements.

### CODE REVIEW

**Comment by MultiDream** [comment]
Reviewed the merge request. Seems like the rewrite was a good way to avoid a for loop, but it doesn't pass all tests.
Assertion error for tests on line 204. Looks like the output doesn't match expectations.

Only one thing I'd note:
There is a reshaping left up to numpy to resolve on runtime. This may have some unexpected consequences if the shape of num_ids is not quite what you expect. That being said, I can't see any reason that would be an issue at this time, so if you feel comfortable with it, don't worry about it. Might be related to the issue, might not be.

**Comment by JHDeerin** [comment]
I believe the failing test is an [xfail test](https://docs.pytest.org/en/reorganize-docs/new-docs/user/xfail.html), not an actual failure; the reshape is actually unnecessary and should be removed with the latest commit.

File: api/algorithm_site_instance.py [hunk 1]
```diff
@@ -347,6 +345,8 @@ def non_nan(array: np.ndarray):
             # Needed since converting NaN to int has undefined behavior
             return array[~np.isnan(array)]
 
+        # alldistance = ids, loci, max_moi**2
```
Comment (danewrye) [inline api/algorithm_site_instance.py]: Is this code unused, and can now be deleted?

File: api/algorithm_site_instance.py [hunk 2]
```diff
@@ -355,9 +355,11 @@ def non_nan(array: np.ndarray):
                 # Double-transpose to multiply across rows, not columns
                 (state.frequencies_RR[1][y, :int(state.frequencies_RR[0][y])]
                 * state.dvect[
+                    # Correction distance matrices are NOT all the same size
                     state.correction_distance_matrix[y][
                         :,
-                        non_nan(state.allrecrf[x, y, :max_MOI**2]).astype(int),
+                        # all_recrf = ids, loci, max_moi**2
```
Comment (danewrye) [inline api/algorithm_site_instance.py]: Same comment as above: this looks like code that has been commented out and can now be deleted.

File: README.md [hunk 3]
```diff
@@ -84,6 +84,7 @@ Georgia Tech Junior Design 2020 project; built in collaboration with Dr. Mateusz
 pip install numpy
 pip install pandas
 pip install scipy
+pip install bottleneck
```
Comment (danewrye) [inline README.md:87]: Cool library! Optimizing the optimized

**Review by danewrye — CHANGES_REQUESTED** [review]
Code looks great, just a couple of extremely minor comment questions. 

When tested, it did not run into any failed tests for me.

File: api/algorithm_site_instance.py (hunk 1 above)
Comment (JHDeerin) [reply api/algorithm_site_instance.py]: This code should already be deleted on the latest commit; I think you reviewed one of the earlier commits accidentally?

Comment (danewrye) [reply api/algorithm_site_instance.py]: Apologies, I will approve then. Everything else looks good.

**Review by danewrye — APPROVED** [review]

**Pull request merged**